Fix pd.merge to preserve ExtensionArrays dtypes

jorisvandenbossche opened the pull request:
Closes https://github.com/pandas-dev/pandas/issues/20743

jorisvandenbossche committed:
Fix pd.merge to preserve ExtensionArrays dtypes

jorisvandenbossche commented on pandas/core/internals.py:
```diff
@@ -5541,7 +5541,7 @@ def concatenate_join_units(join_units, concat_axis, copy):
         # Only one block, nothing to concatenate.
         concat_values = to_concat[0]
-        if copy and concat_values.base is not None:
+        if copy and getattr(concat_values, 'base', 1) is not None:
```
Is this a hack, would like to have a better solution.

(related to discussion earlier this day in https://github.com/pandas-dev/pandas/pull/20721#discussion_r182716666 about deprecating Index.base)

jreback replied:
you should only check base if concat_values is an ndarray

jorisvandenbossche replied:
@jreback updated

jorisvandenbossche commented:
Note to self: 1 actual failing test (rest are network failures):

```
___________________________ TestReshaping.test_merge ___________________________
[gw1] linux -- Python 3.6.5 /home/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.extension.json.test_json.TestReshaping object at 0x7fb92d208dd8>
data = JSONArary([{'D': 91, 'k': 84, 'L': 34, 'm': 63, 'C': 8, 'O': 1, 'R': 5}, {'T': 49, 'o': 46}, {'g': 81, 'O': 8, 'p': 54...': 32, 'p': 0, 'J': 89, 'd': 32, 's': 22}, {'W': 70, 'Y': 68, 'D': 87, 'k': 52}, {'F': 29, 'h': 10, 'S': 41, 'y': 24}])
na_value = {}
    def test_merge(self, data, na_value):
    
        df1 = pd.DataFrame({'int1': [1, 2, 3], 'key': [0, 1, 2],
                            'ext': data[:3]})
        df2 = pd.DataFrame({'int2': [1, 2, 3, 4], 'key': [0, 0, 1, 3]})
    
        res = pd.merge(df1, df2)
        exp = pd.DataFrame(
            {'int1': [1, 1, 2], 'int2': [1, 2, 3], 'key': [0, 0, 1],
             'ext': data._constructor_from_sequence(
                 [data[0], data[0], data[1]])})
>       self.assert_frame_equal(res, exp[['ext', 'int1', 'key', 'int2']])
pandas/tests/extension/base/reshaping.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

E       AssertionError: DataFrame.columns are different
E       
E       DataFrame.columns values are different (75.0 %)
E       [left]:  Index(['int1', 'key', 'ext', 'int2'], dtype='object')
E       [right]: Index(['ext', 'int1', 'key', 'int2'], dtype='object')
```

Not seeing this locally, and also not failing for Decimal, so the column order seems to depend on the extension dtype? (which is a bit strange)

jreback commented on pandas/tests/extension/base/reshaping.py:
```diff
@@ -95,3 +95,24 @@ def test_set_frame_overwrite_object(self, data):
         assert df.dtypes['A'] == data.dtype
+
+    def test_merge(self, data, na_value):
+
```
issue number

TomAugspurger commented:
I can reproduce your failure locally. Looking into it now.

TomAugspurger commented:
Ah, bug in `BaseDecimal.assert_frame_equal` not checking order of all the columns. It only considered extension & non extension columns independently.

TomAugspurger committed:
Fixed order.

The Decimal base class failed to check the global order of columns. Fixed
that as well.

TomAugspurger committed:
Added issue number

TomAugspurger commented:
Pushed a commit fixing the Decimal tests base class and adjust the expected order. Hopefully this will pass things.

jorisvandenbossche committed:
copy: check for arrays

jorisvandenbossche committed:
change order in tests for python < 3.6

jorisvandenbossche commented:
Thanks for the fix! The order was indeed the culprit, but changed it again to be alphabetically in the original test frame, so it behaves the same for python < 3.6

jorisvandenbossche commented:
For Travis, the remaining failure is an s3 one, so this seems to be passing now.

jreback commented on pandas/tests/extension/base/reshaping.py:
```diff
@@ -95,3 +95,24 @@ def test_set_frame_overwrite_object(self, data):
         df['A'] = data
         assert df.dtypes['A'] == data.dtype
+
+    def test_merge(self, data, na_value):
```
you prob should test with with the ``how=join_type`` fixture.

jreback replied:
ok, its worth doing way more tests than a single usecase. but ok here I guess.
